File: crust_env/legacy_c/src/string_ops.c
#include "string_ops.h"

/* Leaf node — no internal dependencies */

size_t str_len(const char *s) {
    size_t len = 0;
    while (s[len] != '\0') {
        len++;
    }
    return len;
}

void to_upper(char *s) {
    for (size_t i = 0; s[i] != '\0'; i++) {
        if (s[i] >= 'a' && s[i] <= 'z') {
            s[i] = s[i] - 32;
        }
    }
}

void to_lower(char *s) {
    for (size_t i = 0; s[i] != '\0'; i++) {
        if (s[i] >= 'A' && s[i] <= 'Z') {
            s[i] = s[i] + 32;
        }
    }
}

int str_equals(const char *a, const char *b) {
    size_t i = 0;
    while (a[i] != '\0' && b[i] != '\0') {
        if (a[i] != b[i]) return 0;
        i++;
    }
    return a[i] == b[i];  /* Both must hit '\0' at same position */
}

void str_copy(char *dest, const char *src) {
    size_t i = 0;
    while (src[i] != '\0') {
        dest[i] = src[i];
        i++;
    }
    dest[i] = '\0';
}

File: crust_env/dummy_workspace/src/string_ops.rs
//! string_ops — String utility operations (leaf node, no dependencies).
//!
//! Rust translation target for: legacy_c/src/string_ops.c
//!
//! Agent must implement idiomatic Rust equivalents.
//! Note: Rust strings are UTF-8, but C legacy code is ASCII-only.
//! Constraints: no `unsafe`, CBO < 3.

/// Returns the number of bytes in the string (UTF-8 length).
pub fn str_len(s: &str) -> usize {
    s.len()
}

/// Returns a new uppercase String (ASCII-only, matching C behavior).
pub fn to_upper(s: &str) -> String {
    s.chars()
        .map(|c| c.to_ascii_uppercase())
        .collect()
}

/// Returns a new lowercase String (ASCII-only, matching C behavior).
pub fn to_lower(s: &str) -> String {
    s.chars()
        .map(|c| c.to_ascii_lowercase())
        .collect()
}

/// Returns true if both strings are equal.
pub fn str_equals(a: &str, b: &str) -> bool {
    a == b
}
